keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,854 @@
+while impacting the fewest connected users on downscale, this allows me to save money by not under-utilizing Nodes,
+and improve user-experience by minimising the number of users impacted by scaling.
+
+Solution:
+
+- I can run my Minecraft server shards in a Deployment.
+- I can use my in-house solution to control how many `replicas` the Deployment has.
+- I can create a REST API (with Java) for `downscalePodPicker` that runs in a Deployment, and answers a request to choose `N` Pods by:
+   1. keeping an in-memory cache of how many users are on each shard (weighting "premium" users higher)
+   2. finding the `N` shards with the lowest user-load:
+       - if multiple shards have the same weighting, return them as "tied pods"
+       - if we find `N` empty shards, we can exit early, and return those as "chosen pods"
+   3. returning these lists of "chosen pods" and "tied pods"
+
+#### Story 3
+
+As a Data Engineer, I want to run an [Apache Spark](https://github.com/apache/spark) cluster and autoscale the number of workers
+while impacting the fewest running tasks on downscale, this allows me to save money by not under-utilizing Nodes,
+and reduce wasted time by minimising how many tasks are impacted by scaling.
+
+Solution:
+
+- (Similar to Story 1)
+
+#### Story 4
+
+As a Site Reliability Engineer, I want to ensure my NodeJS website maintains regional distribution when downscaling the
+number of replicas, this allows me to ensure uptime when a region experiences an outage.
+
+Solution:
+
+- I can run my NodeJS application in a Deployment.
+- I can use a HorizontalPodAutoscaler with CPU metrics to control how many `replicas` the Deployment has.
+- I can create a REST API (with TypeScript) for `downscalePodPicker` that runs in a Deployment, and answers a request to choose `N` Pods by:
+   1. keeping track of which region each Pod is in
+   2. finding `N` Pods that we can remove without violating the regional distribution requirements:
+       - if there are multiple acceptable options, we could choose the `N` Pods which are doing the least work
+   1. return these Pods as the "chosen pods"
+
+### Notes/Constraints/Caveats (Optional)
+
+- The `downscalePodPicker` APIs are written by users to suite their specific needs, and are not part of Kubernetes.
+- There are very few constraints on what a `downscalePodPicker` APIs can be, for example:
+   - they may be fully external to the Kubernetes cluster (e.g. running on an external metrics system)
+   - they may be part of the application which it controls (e.g. a REST API exposed by the application itself)
+   - they may be an adjacent Deployment running on the same Kubernetes cluster (e.g. a Python flask app)
+
+### Risks and Mitigations
+
+- A slow `downscalePodPicker` API may increase the time to downscale a ReplicaSet
+   - MITIGATION: we will provide a `timeoutSeconds` config, if the API has not successfully returned within this time, 
```

Comment:
Is there really any danger in waiting as long as the user wants, (it only impacts the ReplicaSet on which it is set)?

(If the user sets a really high timeout, they probably want to be really careful about which Pods they are killing, so blocking the downscale until it returns is probably what they want.)

Do other timeouts, like those from [Probe v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#probe-v1-core) have an upper limit?

I am not even sure what a sensible hard limit is.

Reply:
The replicaset-controller has to operate on all replicasets. If syncing one replicaset is slow, it potentially takes longer to sync other replicasets.

I don't know for sure if the Probe API has an upper limit. The advantage there is that only one node is affected, not the entire cluster.

A sensible hard limit should probably be correlated with SLOs https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md

Reply:
@alculquicondor do you have a specific idea in mind?

I think it's important to allow a reasonably high timeout (to allow as many use cases as possible).

To prevent one slow API blocking the replicaset-controller from managing all the other ReplicaSets, can't we use an asynchronous call to the Pod-Picker API, and defer interpreting the results until it returns, while continuing to reconcile the other ReplicaSets? (I am not sure how significant of a change is required in the controller to achieve this)

Reply:
Alternatively, the Pod-Picker API could be a "polling" approach, in which the replicaset-controller submits a "job" to the Pod-Picker API, and then polls for the result on the next reconciliation loop (timing out if the defined threshold is exceeded).

But this make the Pod-Picker API contract more complex for end-users to implement.